### PR TITLE
SCTP support implementation for Kubernetes

### LINKED ISSUE #44485: Kubernetes service can't support SCTP protocol
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
No.

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
service sctp 
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
FEATURE REQUEST

**Kubernetes version** (use `kubectl version`):
# kubectl version
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1", GitCommit:"b0b7a323cc5a4a2019b2e9520c21c7830b7f708e", GitTreeState:"clean", BuildDate:"2017-04-03T20:44:38Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:24:30Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**: hardware configuration
- **OS** (e.g. from /etc/os-release):
# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

- **Kernel** (e.g. `uname -a`):
# uname -a
Linux foss-ssc-3 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**:
kubeadm
- **Others**:

**What happened**:
Kubernetes service can't support SCTP as protocol:
Error from server (Invalid): error when creating "tonyaw-svc.yaml": Service "tonyaw-rf" is invalid: spec.ports[0].protocol: Unsupported value: "SCTP": supported values: TCP, UDP

**What you expected to happen**:
protocol should support TCP, UDP and SCTP

**How to reproduce it** (as minimally and precisely as possible):
kind: Service
apiVersion: v1
metadata:
  name: tonyaw-rf
spec:
  clusterIP: None
  selector:
    app: myapp
  ports:
    - name: app
      protocol: SCTP
      port: 7890
      targetPort: 7890

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by janosi**
**What this PR does / why we need it**: This PR adds SCTP support to Kubernetes, including Service, Endpoint, and NetworkPolicy.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #44485

**Special notes for your reviewer**:

**Release note**:

```release-note

SCTP is now supported as additional protocol (alpha) alongside TCP and UDP in Pod, Service, Endpoint, and NetworkPolicy.  

```

### CODE REVIEW

**Comment by thockin** [comment]
I know precious little about SCTP - I have no hands-on experience.  @kubernetes/sig-network-feature-requests I do know, for example that GCP's load-balancer only supports SCTP in single-endpoint mode:  https://cloud.google.com/compute/docs/load-balancing/network/forwarding-rules

So at a minimum, all of the cloud implementations need to be checked for compatibility and disabled when not compatible.

This should probably start as a proposal to sig-network to gather context (like the above GCP point) and then become a KEP for broader review.

**Comment by janosi** [comment]
@thockin 
I have checked now the EnsureLoadBalancer implementations of the different cloud providers (pkg/cloudprovider/providers)
AWS: it accepts only "TCP" as Protocol. Otherwise it returns with error. https://github.com/kubernetes/kubernetes/blob/52603a78abf958c2b3f4b045a21b880dd7606033/pkg/cloudprovider/providers/aws/aws.go#L3249
Azure: it accepts only "TCP" or "UDP". Otherwise it returns with error. https://github.com/kubernetes/kubernetes/blob/52603a78abf958c2b3f4b045a21b880dd7606033/pkg/cloudprovider/providers/azure/azure_standard.go#L176
CloudStack: it accepts only "TCP" or "UDP". Otherwise it returns with error. https://github.com/kubernetes/kubernetes/blob/52603a78abf958c2b3f4b045a21b880dd7606033/pkg/cloudprovider/providers/cloudstack/cloudstack_loadbalancer.go#L455
GCE: it allows only "TCP" or "UDP" for external LB. Otherwise it returns with error. For internal LB there is no explicit protocol check. https://github.com/kubernetes/kubernetes/blob/52603a78abf958c2b3f4b045a21b880dd7606033/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L750
OpenStack: my understanding is, that OpenStack supports SCTP, so I enabled SCTP in the  OpenStack cloudprovider code with this PR
oVirt, Photon, vSphere: these require the registration of a "cloud provider" and it is then up to the "LoadBalancer" implementation of the cloud provider how it handles the different protocols.


pkg/controller/service/service_controller.go also have a nice warning at the point where it calls the EnsureLoadBalancer() function of the relevant cloud provider:
"	// - Not all cloud providers support all protocols and the next step is expected to return
	//   an error for unsupported protocols"
https://github.com/kubernetes/kubernetes/blob/52603a78abf958c2b3f4b045a21b880dd7606033/pkg/controller/service/service_controller.go#L372

As I can see all the supported cloud providers follows this principle, and they handle unsupported protocol values in their code.

**Comment by MaximProshin** [comment]
LKSCTP is not loaded by default in Linux OS that allows developers to run user-land SCTP in Linux. With your change LKSCTP will automatically be loaded and it will intercept all incoming SCTP packets which will lead to aborting of SCTP associations. In other words, your change will not allow developers to run user-land SCTP in K8s cluster.

**Comment by janosi** [comment]
@MaximProshin Is there the same effect when those are in different network namespaces? I.e. when the user space SCTP implementation is used inside a pod that does not use the host network namespace?

File: pkg/kubelet/dockershim/network/hostport/hostport.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ import (
 
 	"k8s.io/api/core/v1"
 	utiliptables "k8s.io/kubernetes/pkg/util/iptables"
+	
+	"github.com/nokia/sctp"
```
Comment (dcbw) [inline pkg/kubelet/dockershim/network/hostport/hostport.go]: I think you need to update the BUILD file in this directory to account for this import addition.  run 'hack/update-bazel.sh' and I think it'll do that for you.

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ import (
 	"time"
 
 	"github.com/golang/glog"
+	"github.com/nokia/sctp"
```
Comment (dcbw) [inline pkg/proxy/iptables/proxier.go]: Same here, bazel update.

File: pkg/proxy/ipvs/proxier.go [hunk 3]
```diff
@@ -1656,6 +1667,18 @@ func openLocalPort(lp *utilproxy.LocalPort) (utilproxy.Closeable, error) {
 			return nil, err
 		}
 		socket = conn
+	case "sctp":
+	  //SCTP is not supported by golang/net, or any other built-in lib,
```
Comment (dcbw) [inline pkg/proxy/ipvs/proxier.go]: mis-indented whitespace; also there should be a single space after the // in comments.

**Comment by janosi** [comment]
@dcbw Thank you for the comments! Updated with hack/update-gofmt.sh and hack/update-bazel.sh. Also did a rebase to the current master.

**Comment by MaximProshin** [comment]
Yes, namespaces don't help to isolate user-land SCTP from LKSCTP so there should be a way to not load/unload LKSCTP at least. If I compare K8s with native Linux OS, I would say the PR is non-backward compatible as previously LKSCTP wasn't loaded by default while with K8s it will be so.

**Comment by janosi** [comment]
@MaximProshin My understanding is that lksctp is not loaded as long as no SCTP Service (with Cluster IP or NodePort) is created, because of the on-demand loading. Once someone creates an SCTP Service the module is loaded, but it is an explicit request from the user. Just like when a user started a native Linux process that wanted to use SCTP.
If my understanding is wrong, please point to the place that triggers lksctp loading right at node startup, i.e. before there is an explicit request for that. Thank you!

**Comment by janosi** [comment]
@thockin I created a KEP, the PR is at https://github.com/kubernetes/community/pull/2276

**Comment by janosi** [comment]
Sure, we can take it to the next SIG Network meeting. Am I allowed to write it on the agenda?

**Comment by danwinship** [comment]
yes

File: pkg/proxy/ipvs/proxier.go [hunk 4]
```diff
@@ -29,6 +29,7 @@ import (
 	"time"
 
 	"github.com/golang/glog"
+	"github.com/nokia/sctp"
```
Comment (matrohon) [inline pkg/proxy/ipvs/proxier.go]: Do you plan to use github.com/ishidawataru/sctp in the final version?

**Comment by MaximProshin** [comment]
@janosi I agree that it was not supported earlier in Linux OS but the difference is that earlier it was up to the customer/engineer to load LKSCTP or use user-land SCTP. With your patch, it will be impossible to run user-land SCTP.
The reason of the problem with LKSCTP vs user-land SCTP is that in your path SCTP socket is opened. If you avoid it then everything will be working. Is it possible to make some option which will prevent to open SCTP socket on a node?

**Comment by matrohon** [comment]
@MaximProshin I'm interested in better understanding the issue you highlight because I use to open sctp sockets from my containers, which leads to sctp LK module loading, as explained by @janosi. Can you review the code and tell us where exactly the patch opens a sctp socket?

**Comment by MaximProshin** [comment]
@matrohon After a brief look at the patch, it seems the following code triggers loading of LKSCTP:

pkg/proxy/iptables/proxier.go:
case "sctp":
                        // There is not any golang/net way to bind or listen on an SCTP socket.
                        // We have to use a 3rd part lib - the same which is used in Docker
                        addr, err := sctp.ResolveSCTPAddr("sctp", net.JoinHostPort(lp.IP, strconv.Itoa(lp.Port)))
                        if err != nil {
                                    return nil, err
                        }
                        conn, err := sctp.ListenSCTP("sctp", addr)
                        if err != nil {
                                    return nil, err
                        }
                        socket = conn

sctp.ListenSCTP function is a part of "github.com/nokia/sctp". It does the following:

sock, err := syscall.Socket(
                        af,
                        syscall.SOCK_STREAM,
                        syscall.IPPROTO_SCTP)

**Comment by KomorkinMikhail** [comment]
It would be great to have support for both LKSCTP and user-land SCTP within this patch. Is it possible to introduce some new boolean parameter ('openServicePort') for services with SCTP Proto which will define whether or not we should open and listen the socket? Thus user will be able to choose which version of SCTP to use.
Yes, it will be only possible to use either LKSCTP or user-land SCTP at a time, but looks like there's no other way to support user-land version since LKSCTP loads up on all nodes during service creation.

**Comment by janosi** [comment]
@KomorkinMikhail Please check the KEP for proposals how we could solve the interworking problem.

**Comment by janosi** [comment]
@thockin Could you please advise how to move this one forward?

**Comment by janosi** [comment]
@dep-approvers @api-approvers @dchen1107 @api-approvers @brendandburns @bowei @sig-cli-maintainers @derekwaynecarr @gmarek @smarterclayton Please check, review, and if acceptable, approve this one.

**Comment by cblecker** [comment]
/hold

@janosi For clarity, this PR isn't ready for approval.

Things that need to happen first:
- The KEP (https://github.com/kubernetes/community/pull/2276) needs to be approved by sig-network, merged, designated as implementable.
- Once that's done, then this PR would need both a technical review, as well as an API-specific review as it modifies our API. This would be done by sig-architecture.
- Only once those two pieces are done do you need to worry about approvers.

**Comment by m1093782566** [comment]
I would prefer small PR for the sake of easily review.

**Comment by janosi** [comment]
/assign @thockin 
The KEP https://github.com/kubernetes/community/pull/2276 is merged now, and approved by @thockin  as "implementable".
@thockin I updated the code according to the KEP, could you please advice how to move forward with the technical review?
@m1093782566 I think, this PR implements the SCTP support as a feature.

**Comment by janosi** [comment]
Hopefully this PR https://github.com/kubernetes/test-infra/pull/9161 solved the GKE test issues on 27th August 2018, so
/retest

**Comment by janosi** [comment]
@thockin Thank you for guiding me through this adventure, for the review, and for the approval!
@MaximProshin @danwinship @dcbw @KomorkinMikhail @matrohon @m1093782566  Thank you for your comments and reviews!

**Comment by justaugustus** [comment]
Linking the features issue for this: https://github.com/kubernetes/features/issues/614

**Comment by yujuhong** [comment]
I don't think this gets enough attention from the sig-node. This adds a new SCTP protocol to CRI and may require changes in the CRI shims.

@kubernetes/sig-node-api-reviews @kubernetes/sig-node-pr-reviews @mrunalp @Random-Liu @resouer @feiskyer

**Comment by feiskyer** [comment]
> On the Kubernetes side we solve this problem so, that we do not start listening on the SCTP ports defined for Servies with ClusterIP or NodePort or externalIP, neither in the case when containers with SCTP HostPort are defined. On this way we avoid the loading of the kernel module due to Kubernetes actions.

@yujuhong Should we skip SCTP port mapping in kuberuntime instead?

**Comment by janosi** [comment]
@feiskyer @yujuhong I wonder whether the same logic could be applied on the CRI side which was selected for the load balancers: k8s itself does not restrict the usage of SCTP as protocol for load balancer reuqests, and it is the task of the cloud provider's load balancer plugin to filter unsupported protocols.
For example, AFAIK, Docker supports SCTP for host port binding.

**Pull request merged**